src/dnscap.c

```diff
@@ -2339,8 +2375,61 @@ network_pkt(const char *descr, my_bpftimeval ts, unsigned pf,
 			abort();
 		}
 	}
-	output(descr, from, to, proto, flags, sport, dport, ts,
-	    pkt_copy, olen, dnspkt, dnslen);
+/*Sample Module*/
+        if (sample == TRUE)
+        {
+                ns_msg dnsmsgSample;
+                ns_initparse(dnspkt,dnslen,&dnsmsgSample);
+                samplePacket *currentQuery;
+                void *ipAddrBufferFrom = &from.u;
+                void *ipAddrBufferTo = &to.u;
+                unsigned *fromBuffer = (unsigned*)ipAddrBufferFrom;
+                unsigned *toBuffer = (unsigned*)ipAddrBufferTo;
+
+                keylen = offsetof(samplePacket,to)      //keylen is used to define which fields of the hash structure are added
+                        + sizeof(*toBuffer)                    //as a compound key. Here, the key is composed of all fields between (and including)
+                        - offsetof(samplePacket,from);  //samplePacket->to and samplePacket->from (from, sport, dport, transaction_id, to)
+
+                if(dns.qr == 0)
+                {
+                        querycount++;
+                        if(querycount % sampleAmount == 0)
+                                {
+                                        currentQuery = malloc(sizeof(*currentQuery));
+                                        memset(currentQuery, 0, sizeof(*currentQuery));
+                                        currentQuery->from = *fromBuffer;
```

Comment:
`samplePacket->from` should contain IP addresses (both IPv4 and v6 would be optimal).

On my system `unsigned` is 32 bits (which is the size of IPv4 adresses), so it should be able to store a full IPv4. I'll change it to `uint_32` so it also works on all systems.
As to IPv6, I did not find out how to handle such big numbers properly. I thought it would be better to take a small portion of the IPv6 addresses than not to work with IPv6 at all.